---
title: "RNA_analysis"
author: "Hypertyz"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
library(dplyr)
library(DESeq2)
library(tximport)
library(EnhancedVolcano)
library(ggpubr)

file_names = c("SRR9665408",
          "SRR9665409",
          "SRR9665410",
          "SRR9665411",
          "SRR9665412",
          "SRR9665413",
          "SRR9665414",
          "SRR9665415",
          "SRR9665416",
          "SRR9665417",
          "SRR9665418",
          "SRR9665419"
)

```

## RNA-seq analysis to detect promoter sequences

```{r load data}
abundance <- read.delim(paste("./sample_reads/", file_names[1], "/abundance.tsv",
                              sep = ""))

abundance <- abundance %>% select(target_id, tpm) %>% rename_with(recode, tpm = file_names[1])



for (i in 2:length(file_names)) {
  tmp <- read.delim(paste("./sample_reads/", file_names[i], "/abundance.tsv",
                          sep = ""))
  
  abundance <- tmp %>% select(target_id, tpm) %>%  rename_with(recode, tpm = file_names[i]) %>% left_join(abundance, by ="target_id")
  
}

```

```{r prepare matrix}
source("y00_functions.R")
metaData <- read_clean_metaData()
metaData <- metaData[-5:-7]

files <- file.path("./sample_reads", file_names, "abundance.h5")

names(files) <- paste0(metaData$Name)

txi.kallisto <- tximport(files, type = "kallisto", txOut = TRUE, countsFromAbundance = "lengthScaledTPM")

ddsTxi <- DESeqDataSetFromTximport(txi.kallisto,
                                   colData = metaData,
                                   design = ~ condition)

```

### DESeq2

```{r DESeq2}
dds <- DESeq(ddsTxi)
res <- results(dds, contrast=c("condition","Fa","C"))

res <- res[order(res$padj),]
summary(res)
head(res)

```

```{r Volcano plot}

theme_set(theme_pubr())

pCutoff = 0.05
FCcutoff = 2.0

res_df <- as.data.frame(res@listData)
g_names <- as.data.frame(res@rownames)

res_df <- cbind(g_names, res_df)
colnames(res_df)[1] <- "Gene_id"

res_df <- res_df %>% filter(padj <0.05 & log2FoldChange > 2.0)
#Interesting genes

# YOR388C
# YNR034W-A
# YOL052C-A
# chosen_genes <- c("YOR388C","YNR034W-A","YOL052C-A")
chosen_genes <- c(res_df$Gene_id)


EnhancedVolcano(res,
                lab = rownames(res),
                selectLab = chosen_genes[0:3],
                x = 'log2FoldChange',
                y = 'pvalue',
                title = "Furfural",
                subtitle = bquote(italic(" ")),
                legendPosition = 'top',
                pCutoff = pCutoff,
                FCcutoff = FCcutoff,
                drawConnectors = TRUE,
                widthConnectors = 1.0,
                ylim = c(0,350),
                xlim = c(-5,8)
                )


```

```{r annotation}
gene_info <- read.csv("./sample_info/gene_info.csv", header=FALSE, row.names=NULL, stringsAsFactors=TRUE)

colnames(gene_info) <- c("Gene_id", "Gene_name", "Loci")

res_df <- res_df %>% left_join(gene_info, by = "Gene_id")

```

```{r promoters}
library("Biostrings")

genome = readDNAStringSet("./sample_info/S288C_reference_sequence_R64-3-1_20210421.fsa")

annot_dict <- c("I"=1,
                "II"=2,
                "III"=3,
                "IV"=4,
                "V"=5,
                "VI"=6,
                "VII"=7,
                "VIII"=8,
                "IX"=9,
                "X"=10,
                "XI"=11,
                "XII"=12,
                "XIII"=13,
                "XIV"=14,
                "XV"=15,
                "XVI"=16,
                "Mito"=17)
nuc_seq <- c()

for (i in 1:length(res_df$Loci)){
  if (grepl( ";", res_df$Loci[i], fixed = TRUE)) {
    nuc_seq_MM <- c()
    for (j in 1:length(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], ";")[[1]])){
      nuc_seq_MM <- append(nuc_seq_MM, paste(j, "-", as.character(genome[[annot_dict[strsplit(as.character(res_df$Loci[i]), " ")[[1]][2]]]][as.numeric(strsplit(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], ";")[[1]][j], "-")[[1]][1]):as.numeric(strsplit(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], ";")[[1]][j], "-")[[1]][2])]), sep = ""))
    }
    nuc_seq_MM <- paste(nuc_seq_MM, collapse = ";")
    nuc_seq <- append(nuc_seq, nuc_seq_MM)
  } else {
    nuc_seq <- append(nuc_seq, as.character(genome[[annot_dict[strsplit(as.character(res_df$Loci[i]), " ")[[1]][2]]]][as.numeric(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], "-")[[1]][1]):as.numeric(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], "-")[[1]][2])]))
  }
}

res_df <- cbind(res_df, nuc_seq)

promo_seq <- c()

for (i in 1:length(res_df$Loci)){
  if (grepl( ";", res_df$Loci[i], fixed = TRUE)) {
    promo_seq_MM <- c()
    for (j in 1:length(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], ";")[[1]])){
      if (as.numeric(strsplit(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], ";")[[1]][j], "-")[[1]][1]) < as.numeric(strsplit(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], ";")[[1]][j], "-")[[1]][2])) {
        promo_seq_MM <- append(promo_seq_MM, paste(j, "-", as.character(genome[[annot_dict[strsplit(as.character(res_df$Loci[i]), " ")[[1]][2]]]][as.numeric(as.numeric(strsplit(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], ";")[[1]][j], "-")[[1]][1])-999):as.numeric(strsplit(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], ";")[[1]][j], "-")[[1]][1])]), sep = ""))
      } else {
        promo_seq_MM <- append(promo_seq_MM, paste(j, "-", as.character(genome[[annot_dict[strsplit(as.character(res_df$Loci[i]), " ")[[1]][2]]]][as.numeric(as.numeric(strsplit(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], ";")[[1]][j], "-")[[1]][1])+999):as.numeric(strsplit(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], ";")[[1]][j], "-")[[1]][1])]), sep = ""))
      }
      }
    promo_seq_MM <- paste(promo_seq_MM, collapse = ";")
    promo_seq <- append(promo_seq, promo_seq_MM)
  } else {
    if (as.numeric(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], "-")[[1]][1]) < as.numeric(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], "-")[[1]][2])) {
      promo_seq <- append(promo_seq, as.character(genome[[annot_dict[strsplit(as.character(res_df$Loci[i]), " ")[[1]][2]]]][as.numeric(as.numeric(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], "-")[[1]][1])-999):as.numeric(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], "-")[[1]][1])]))
    } else {
      promo_seq <- append(promo_seq, as.character(genome[[annot_dict[strsplit(as.character(res_df$Loci[i]), " ")[[1]][2]]]][as.numeric(as.numeric(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], "-")[[1]][1])+999):as.numeric(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], "-")[[1]][1])]))
    }
  }
}

res_df <- cbind(res_df, promo_seq)

strand <- c()

for (i in 1:length(res_df$Loci)){
  if (grepl( ";", res_df$Loci[i], fixed = TRUE)) {
    strand_MM <- c()
    for (j in 1:length(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], ";")[[1]])){
      if (as.numeric(strsplit(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], ";")[[1]][j], "-")[[1]][1]) < as.numeric(strsplit(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], ";")[[1]][j], "-")[[1]][2])) {
        strand_MM <- append(strand_MM, "+")
      } else {
        strand_MM <- append(strand_MM, "-")
      }
      }
    strand_MM <- paste(strand_MM, collapse = ";")
    strand <- append(strand, strand_MM)
  } else {
    if (as.numeric(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], "-")[[1]][1]) < as.numeric(strsplit(strsplit(as.character(res_df$Loci[i]), " ")[[1]][4], "-")[[1]][2])) {
      strand <- append(strand, "+")
    } else {
      strand <- append(strand, "-")
    }
  }
}

res_df <- cbind(res_df, strand)

# prom_can <- res_df %>% filter(strand == "+" | strand == "-") %>% select(promo_seq)
prom_can <- res_df %>% filter(strand == "+") %>% select(promo_seq)

```

```{r save}
write.table(prom_can, "./R_outputs/promoters.csv", append = FALSE, sep = "\n",
            row.names = FALSE, col.names = FALSE, quote = FALSE)

write.table(res_df, "./R_outputs/DEseq_result.csv", append = FALSE, sep = ",",
            row.names = FALSE, col.names = FALSE, quote = FALSE)

```
