---
title: "RNA_analysis"
author: "Hypertyz"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
library(dplyr)
library(DESeq2)
library(tximport)
library(EnhancedVolcano)
library(ggpubr)
library(readr)
library(seqinr)
library(MotifDb)
library(seqLogo)
library(motifStack)
library(Biostrings)
library(GenomicFeatures)

file_names = c("SRR9665408",
          "SRR9665409",
          "SRR9665410",
          "SRR9665411",
          "SRR9665412",
          "SRR9665413",
          "SRR9665414",
          "SRR9665415",
          "SRR9665416",
          "SRR9665417",
          "SRR9665418",
          "SRR9665419"
)

```

## RNA-seq analysis to detect promoter sequences

```{r load data}
abundance <- read.delim(paste("./sample_reads/", file_names[1], "/abundance.tsv",
                              sep = ""))

abundance <- abundance %>% dplyr::select(target_id, tpm) %>% rename_with(recode, tpm = file_names[1])



for (i in 2:length(file_names)) {
  tmp <- read.delim(paste("./sample_reads/", file_names[i], "/abundance.tsv",
                          sep = ""))
  
  abundance <- tmp %>% dplyr::select(target_id, tpm) %>%  rename_with(recode, tpm = file_names[i]) %>% left_join(abundance, by ="target_id")
  
}

```

```{r prepare matrix}
source("y00_functions.R")
metaData <- read_clean_metaData()
metaData <- metaData[-5:-7]

files <- file.path("./sample_reads", file_names, "abundance.h5")

names(files) <- paste0(metaData$Name)

txi.kallisto <- tximport(files, type = "kallisto", txOut = TRUE, countsFromAbundance = "lengthScaledTPM")

ddsTxi <- DESeqDataSetFromTximport(txi.kallisto,
                                   colData = metaData,
                                   design = ~ condition)

```

## DESeq2

```{r DESeq2}
dds <- DESeq(ddsTxi)
res <- results(dds, contrast=c("condition","Fa","C"))

res <- res[order(res$padj),]
summary(res)
head(res)

```

```{r Volcano plot}

theme_set(theme_pubr())

pCutoff = 0.05
FCcutoff = 2.0

res_df <- as.data.frame(res@listData)
g_names <- as.data.frame(res@rownames)

res_df <- cbind(g_names, res_df)
colnames(res_df)[1] <- "Gene_id"

res_df_sig <- res_df %>% dplyr::filter(padj <0.05)
#Interesting genes

# YOR388C
# YNR034W-A
# YOL052C-A
# chosen_genes <- c("YOR388C","YNR034W-A","YOL052C-A")
chosen_genes <- c(res_df_sig$Gene_id)

EnhancedVolcano(res,
                lab = rownames(res),
                selectLab = chosen_genes[0:3],
                x = 'log2FoldChange',
                y = 'padj',
                title = "Furfural",
                subtitle = bquote(italic(" ")),
                legendPosition = 'top',
                pCutoff = pCutoff,
                FCcutoff = FCcutoff,
                drawConnectors = TRUE,
                widthConnectors = 1.0,
                ylim = c(0,350),
                xlim = c(-5,8)
                )


```

```{r annotation}
res_df_sig <- read_delim("./R_outputs/res_DEGs_up_fur.csv", 
    delim = ",", escape_double = FALSE, trim_ws = TRUE)
colnames(res_df_sig)[1] <- "Gene_id"


gene_info <- read.csv("./sample_info/gene_info.csv", header=FALSE, row.names=NULL, stringsAsFactors=TRUE)

colnames(gene_info) <- c("Gene_id", "Gene_name", "Loci")

res_df_sig <- res_df_sig %>% left_join(gene_info, by = "Gene_id")

```
## Enrichment analysis
```{r enrichment}
library(clusterProfiler)

# KEGG pathway enrichment for top 1000 DEGs
pathEnrich <- enrichKEGG(gene = res_df_sig$Gene_id,
                         organism = "sce", # Yeast
                         keyType = "kegg", # KEGG database
                         pvalueCutoff = 0.05,
                         pAdjustMethod = "BH",
                         universe = res_df$Gene_id)

# Store the significantly enriched pathways (P < 0.05)
signif_pathEnrich <- pathEnrich@result[pathEnrich@result$p.adjust < 0.05, ]

##########

# Show the 10 most significant pathways as a dot plot
dotplot(pathEnrich, x = "Count", showCategory = 10, orderBy = "Count")

##########

# 3.2 Functional Class Scoring: Gene Set Enrichment Analysis

# Create a list of logFCs with their corresponding genes as names
logFC <- res_df$log2FoldChange
names(logFC) <- res_df$Gene_id

# Order the list in decreasing order
logFC <- sort(logFC, decreasing = TRUE)

# Perform GSEA using KEGG pathways
gseaKEGG <- gseKEGG(geneList = logFC,
                    organism = "sce",
                    nPerm = 1000, # Number of permutations
                    pvalueCutoff = 0.05)

# Extract the GSEA results
gseaKEGG_results <- gseaKEGG@result

# Plot the GSEA plot for a single enriched pathway
gseaplot(gseaKEGG, geneSetID = "sce01200")

##########

# 3.3 Pathway Topology: pathview

# Install and load the required package
library(pathview)

# Run pathview for a particular pathway ID
# pathview(gene.data = logFC,
#          pathway.id = "sce01200",
#          low = "red", mid = "grey", high = "green",
#          species = "sce",
#          limit = c(-10, 10),
#          gene.idtype = "KEGG",
#          kegg.dir = "./R_outputs/KEGG/")
```
```{r save}
write.csv(signif_pathEnrich, file = "./R_outputs/KEGG/signif_pathEnrich.csv",
          row.names = FALSE, quote = FALSE)

```


## Extract promoters

```{r promoters}
library("Biostrings")

genome = readDNAStringSet("./sample_info/S288C_reference_sequence_R64-3-1_20210421.fsa")

annot_dict <- c("I"=1,
                "II"=2,
                "III"=3,
                "IV"=4,
                "V"=5,
                "VI"=6,
                "VII"=7,
                "VIII"=8,
                "IX"=9,
                "X"=10,
                "XI"=11,
                "XII"=12,
                "XIII"=13,
                "XIV"=14,
                "XV"=15,
                "XVI"=16,
                "Mito"=17)
nuc_seq <- c()

for (i in 1:length(res_df_sig$Loci)){
  if (grepl( ";", res_df_sig$Loci[i], fixed = TRUE)) {
    nuc_seq_MM <- c()
    for (j in 1:length(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], ";")[[1]])){
      nuc_seq_MM <- append(nuc_seq_MM, paste(j, "-", as.character(genome[[annot_dict[strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][2]]]][as.numeric(strsplit(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], ";")[[1]][j], "-")[[1]][1]):as.numeric(strsplit(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], ";")[[1]][j], "-")[[1]][2])]), sep = ""))
    }
    nuc_seq_MM <- paste(nuc_seq_MM, collapse = ";")
    nuc_seq <- append(nuc_seq, nuc_seq_MM)
  } else {
    nuc_seq <- append(nuc_seq, as.character(genome[[annot_dict[strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][2]]]][as.numeric(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], "-")[[1]][1]):as.numeric(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], "-")[[1]][2])]))
  }
}

res_df_sig <- cbind(res_df_sig, nuc_seq)

promo_seq <- c()

for (i in 1:length(res_df_sig$Loci)){
  if (grepl( ";", res_df_sig$Loci[i], fixed = TRUE)) {
    promo_seq_MM <- c()
    for (j in 1:length(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], ";")[[1]])){
      if (as.numeric(strsplit(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], ";")[[1]][j], "-")[[1]][1]) < as.numeric(strsplit(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], ";")[[1]][j], "-")[[1]][2])) {
        promo_seq_MM <- append(promo_seq_MM, paste(j, "-", as.character(genome[[annot_dict[strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][2]]]][as.numeric(as.numeric(strsplit(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], ";")[[1]][j], "-")[[1]][1])-1000):as.numeric(strsplit(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], ";")[[1]][j], "-")[[1]][1])]), sep = ""))
      } else {
        promo_seq_MM <- append(promo_seq_MM, paste(j, "-", as.character(genome[[annot_dict[strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][2]]]][as.numeric(as.numeric(strsplit(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], ";")[[1]][j], "-")[[1]][1])+1000):as.numeric(strsplit(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], ";")[[1]][j], "-")[[1]][1])]), sep = ""))
      }
      }
    promo_seq_MM <- paste(promo_seq_MM, collapse = ";")
    promo_seq <- append(promo_seq, promo_seq_MM)
  } else {
    if (as.numeric(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], "-")[[1]][1]) < as.numeric(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], "-")[[1]][2])) {
      promo_seq <- append(promo_seq, as.character(genome[[annot_dict[strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][2]]]][as.numeric(as.numeric(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], "-")[[1]][1])-1000):as.numeric(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], "-")[[1]][1])]))
    } else {
      promo_seq <- append(promo_seq, as.character(genome[[annot_dict[strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][2]]]][as.numeric(as.numeric(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], "-")[[1]][1])+1000):as.numeric(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], "-")[[1]][1])]))
    }
  }
}

res_df_sig <- cbind(res_df_sig, promo_seq)

strand <- c()

for (i in 1:length(res_df_sig$Loci)){
  if (grepl( ";", res_df_sig$Loci[i], fixed = TRUE)) {
    strand_MM <- c()
    for (j in 1:length(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], ";")[[1]])){
      if (as.numeric(strsplit(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], ";")[[1]][j], "-")[[1]][1]) < as.numeric(strsplit(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], ";")[[1]][j], "-")[[1]][2])) {
        strand_MM <- append(strand_MM, "+")
      } else {
        strand_MM <- append(strand_MM, "-")
      }
      }
    strand_MM <- paste(strand_MM, collapse = ";")
    strand <- append(strand, strand_MM)
  } else {
    if (as.numeric(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], "-")[[1]][1]) < as.numeric(strsplit(strsplit(as.character(res_df_sig$Loci[i]), " ")[[1]][4], "-")[[1]][2])) {
      strand <- append(strand, "+")
    } else {
      strand <- append(strand, "-")
    }
  }
}

res_df_sig <- cbind(res_df_sig, strand)

neg_strand_promo_comp <- res_df_sig %>% dplyr::filter(strand == "-") %>% mutate(promo_seq=replace(promo_seq, strand=="-", chartr("ATGC","TACG",promo_seq)))

pos_strand_promo <- res_df_sig %>% dplyr::filter(strand == "+")

promo_metadata <- bind_rows(neg_strand_promo_comp, pos_strand_promo) %>% arrange(desc(log2FoldChange))

prom_can <- promo_metadata %>% dplyr::select(promo_seq)

```

```{r save}
write.table(prom_can, "./R_outputs/promoters.csv", append = FALSE, sep = "\n",
            row.names = FALSE, col.names = FALSE, quote = FALSE)

write.table(res_df_sig, "./R_outputs/DEseq_result.tsv", append = FALSE, sep = "\t",
            row.names = FALSE, col.names = TRUE, quote = FALSE)

```

## Find DNA-binding sites

```{r extract DNA-binding sites using Motif DB}
points <- promo_metadata$log2FoldChange

clusters <- kmeans(points, centers = 4)

promoters <- promo_metadata %>% bind_cols(clusters$cluster) %>% dplyr::rename(cluster = ...12)

get_DNABS <- function(promoter){
  
  BS <- c()
  
  matrices.yeast <- query(MotifDb, c("Scerevisiae", "JASPAR_CORE"))
  
  for (i in 1:length(matrices.yeast)){
    matrices.yeast.single <- matrices.yeast[[i]]
    matrices.yeast.single <- round(100 * matrices.yeast.single)
    match <- matchPWM(matrices.yeast.single, unlist(promoter)[[1]], "95%")
    if (length(match@ranges@width) != 0){
      BS <- append(names(matrices.yeast[i]), BS)
      }
  }
  return(BS)
}

BSs <- list()

for (i in promoters$cluster[1]){
  promoters_cluster <- as.list(promoters %>% dplyr::filter(cluster == i) %>% dplyr::select(promo_seq))
  promoters_cluster_name <- promoters %>% dplyr::filter(cluster == i) %>% dplyr::select(Gene_name)
  BS <- lapply(promoters_cluster$promo_seq, get_DNABS)
  BSs <- append(list(BS), BSs)
  }

consensus_bs <- c()
for (j in 1:length(BSs[[1]][[1]])) {
  if (BSs[[1]][[1]][j] %in% BSs[[1]][[2]] & BSs[[1]][[1]][j] %in% BSs[[1]][[3]]){
    consensus_bs <- append(BSs[[1]][[1]][j], consensus_bs)
  }
}

seqlogo_plot <- c()

for (i in stringr::str_split(consensus_bs, "-")){
  seqlogo_plot <- append(new("pfm", mat=query(MotifDb, c( "JASPAR_CORE", i[4]))[[1]], name=i[3]), seqlogo_plot)
}

plotMotifLogoStack(DNAmotifAlignment(seqlogo_plot))


```

```{r extract DNA-binding sites based on MAFFT alignment}

write.fasta(as.list(promoters_cluster$promo_seq), as.character(promoters_cluster_name$Gene_name), "./R_outputs/MAFFT/mafft_promo.fasta") 

### Run MAFTT and get reuslt from Jalview

# Consensus
consensus <- read.fasta("./R_outputs/Jalview_output/consensus.fasta", as.string = TRUE)

consensus <- as.character(consensus$Consensus[1])

# Occupancy
occupancy <- read_csv("./R_outputs/Jalview_output/occupancy.csv", 
    col_names = FALSE)

sequence = c()

for (i in 1:nchar(consensus)){
  sequence <- append(sequence, substr(consensus, i, i))
}

sequence <- as.data.frame(sequence)

occupancy <- t(occupancy)
end <- length(occupancy)-1
occupancy <- as.data.frame(occupancy[2:end])

# Alignments
alignments <- read.fasta("./R_outputs/MAFFT/mafft_promo_aligned.fasta", as.string = TRUE)

test_seqs <- as.data.frame(c(1:nchar(consensus)))

for (a in alignments) {
  test_seq <- c()
  s <- a[1]
  for (i in 1:nchar(s)){
    test_seq <- append(test_seq, substr(s, i, i))
  }
  test_seq <- as.data.frame(test_seq)
  test_seqs <- cbind(test_seqs, test_seq)
}

test_seqs <- test_seqs[-1]
colnames(test_seqs) <- as.character(promo_metadata$Gene_name[1:3])

binding_sites <- bind_cols(sequence, test_seqs, occupancy)
colnames(binding_sites)[ncol(binding_sites)] <- "occupancy"

# Binding sites
binding_sites <- binding_sites %>% mutate(sequence=replace(sequence, occupancy < 3, "+"))

idx <- 1
binding_site_seqs <- list()
binding_site <- c()

for (n in binding_sites$sequence){
  if (n == "+"){
    if(length(binding_site) == 0) {
      idx <- idx + 1
    } else {
       binding_site_seqs <- append(binding_site_seqs, list(binding_site))
       binding_site <- c()
       idx <- idx + 1
      }
  } else {
    binding_site <- append(binding_site, idx)
    idx <- idx + 1
  }
  
}

binding_site_seqs_filtered <- list()

for (i in binding_site_seqs){
  if (length(i) >= 6){
    binding_site_seqs_filtered <- append(binding_site_seqs_filtered, list(i))
    }
}

putative_TFBS <- list()
seq_n <- 1

for (i in binding_site_seqs_filtered){
  start <- i[1]
  end <- i[length(i)]
  for (j in test_seqs[start:end,]){
    putative_TFBS <- append(putative_TFBS, list(paste(j, collapse = '')))
  }
  write.fasta(putative_TFBS, colnames(test_seqs),
              paste("./R_outputs/Putative_TFBS/putative_TFBS_", seq_n, ".fasta", sep = ""))
  putative_TFBS <- list()
  seq_n <- seq_n + 1 
}

# PFM - Trine look!
putative_TFBS1 <- read.fasta("./R_outputs/Putative_TFBS/putative_TFBS_1.fasta", as.string = T)
pfm <- c()

for (s in putative_TFBS1) {
  pfm <- append(pfm, s)
}

letterMatrix <- function(input){
  # Ensure kmers are the same length characters(ggseqlogo)
  # First, ensure that the length of the input base sequence is consistent
  seq.len = sapply(input, nchar) # Calculate the number of bases in each sequence
  num_pos = seq.len[1] # Number of bases in the first sequence
  if(! all(seq.len == num_pos)) { # The number of bases of all sequences must be the same. If not, an error will be reported
    stop('Sequences in alignment must have identical lengths')
  }

  # Construct matrix of letters(ggseqlogo)
  # Next, build a matrix, each element is a base
  split = unlist( sapply(input, function(seq){strsplit(seq, '')}) ) # strsplit can cut strings into individual characters

  t( matrix(split, seq.len, length(split)/num_pos) )
}

make_ppm <- function(seqs, ppm=TRUE, seq_type="dna") {
  # seqs: A vector of strings, each string is a DNA or RNA sequence
  # ppm: Whether to return PPM, default is PPM, else return PFM
  # seq_type: Sequence type, can be "dna" of "rna"

  letter_mat = letterMatrix(seqs) # Construct a base matrix, each row is a sequence, and each column is the base position

  # Get namespace(ggseqlogo)
  if(seq_type == "dna") {
    namespace = c("A", "T", "G", "C") 
  } else if (seq_type == "rna" ) {
    namespace = c("A", "U", "G", "C") 
  } else {
    stop('Wrong seq_type! Must be one of "dna" and "rna".')
  }

  # Construct PWM(ggseqlogo)
  pfm_mat = apply(letter_mat, 2, function(pos.data){ # The second parameter of apply is 2, which indicates the operation on the column
    # Get frequencies (ggseqlogo)
    t = table(pos.data) # Calculate the frequency of four bases at this position
    # Match to aa(ggseqlogo)
    ind = match(namespace, names(t)) # 
    # Create column(ggseqlogo)
    col = t[ind] # 
    col[is.na(col)] = 0
    names(col) = namespace

    if(ppm) { # If PPM is returned, divide the base frequency by the total number of bases in this column
      col = col / sum(col)      
    }
    col # Function return value col
  })

  num_pos = nchar(seqs[1])
  colnames(pfm_mat) = 1:num_pos
  pfm_mat

}

pwm <- make_ppm(toupper(pfm))
seqLogo(pwm)

matrices.yeast <- query(MotifDb, c("Scerevisiae", "JASPAR_CORE"))


for (k in binding_site_seqs_filtered){
  start <- k[1]
  end <- k[length(k)]
  for (i in 1:length(matrices.yeast)){
    matrices.yeast.single <- matrices.yeast[[i]]
    matrices.yeast.single <- round(100 * matrices.yeast.single)
    match <- matchPWM(matrices.yeast.single, paste(binding_sites$sequence[start:end], collapse = ''), "95%")
    if (length(match@ranges@width) != 0){
      print(paste(binding_sites$sequence[start:end], collapse = ''))
      print(names(matrices.yeast[i]))
      consensus_bs <- append(names(matrices.yeast[i]), consensus_bs)
    }
  }
}

write.table(as.data.frame(consensus_bs) %>% dplyr::count(consensus_bs), file = "./R_outputs/PuTF.tsv", sep = "\t", quote = F, row.names = F)

```